library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/async/RecordedDataQueueHandler.kt

```diff
@@ -0,0 +1,177 @@
+import com.datadog.android.sessionreplay.internal.processor.RumContextDataHandler
+import com.datadog.android.sessionreplay.internal.recorder.SystemInformation
+import com.datadog.android.sessionreplay.internal.utils.TimeProvider
+import com.datadog.android.sessionreplay.model.MobileSegment
+import java.lang.ClassCastException
+import java.lang.NullPointerException
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.RejectedExecutionException
+
+/**
+ * Responsible for storing the Snapshot and Interaction events in a queue.
+ * Allows for asynchronous enrichment, which still preserving the event order.
+ * The items are added to the queue from the main thread and processed on a background thread.
+ */
+internal class RecordedDataQueueHandler(
+    private val processor: RecordedDataProcessor,
+    private val rumContextDataHandler: RumContextDataHandler,
+    private val executorService: ExecutorService,
+    private val timeProvider: TimeProvider
+) {
+
+    // region internal
+    internal var recordedDataQueue = ConcurrentLinkedQueue<RecordedDataQueueItem>()
+
+    @MainThread
+    internal fun addTouchEventItem(
+        pointerInteractions: List<MobileSegment.MobileRecord>
+    ): TouchEventRecordedDataQueueItem? {
+        val rumContextData = rumContextDataHandler.createRumContextData()
+            ?: return null
+
+        val item = TouchEventRecordedDataQueueItem(
+            timestamp = rumContextData.timestamp,
+            prevRumContext = rumContextData.prevRumContext,
+            newRumContext = rumContextData.newRumContext
+        )
+
+        item.touchData = pointerInteractions
+
+        insertIntoRecordedDataQueue(item)
+
+        return item
+    }
+
+    @MainThread
+    internal fun addSnapshotItem(
+        systemInformation: SystemInformation
+    ): SnapshotRecordedDataQueueItem? {
+        val rumContextData = rumContextDataHandler.createRumContextData()
+            ?: return null
+
+        val item = SnapshotRecordedDataQueueItem(
+            timestamp = rumContextData.timestamp,
+            prevRumContext = rumContextData.prevRumContext,
+            newRumContext = rumContextData.newRumContext,
+            systemInformation = systemInformation
+        )
+
+        item.systemInformation = systemInformation
+
+        insertIntoRecordedDataQueue(item)
+
+        return item
+    }
+
+    /**
+     * Goes through the queue one item at a time for as long as there are items in the queue.
+     * If an item is ready to be consumed, it is processed.
+     * If an invalid item is encountered, it is removed (invalid items are possible
+     * for example if a snapshot failed to traverse the tree).
+     * If neither of the previous conditions occurs, the loop breaks.
+     */
+    @MainThread
+    internal fun tryToConsumeItems() {
+        // no need to create a thread if the queue is empty
+        if (recordedDataQueue.isEmpty()) {
```

Comment:
Ok, you can ignore this comment. I checked the implementation in the Android codebase and it seems we are good here, because `isEmpty` implementation for the `ConcurrentLinkedQueue` [relies on the `first/peek` call](https://cs.android.com/android/platform/superproject/+/master:libcore/ojluni/src/main/java/java/util/concurrent/ConcurrentLinkedQueue.java;l=446;drc=260545b78f2f71d32f1208c90f77220729591f7b).